### ISSUE TITLE
SC2154 fails to trigger when assigning function `local globalvar="$globalvar"` (SC2269 also not triggered)

### ISSUE THREAD

**Issue by iconoclasthero** [koalaman/shellcheck#3453]
#### For bugs with existing features

- Rule Id (if any, e.g. SC1000): SC2154
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```bash
#!/bin/bash
f(){
  ## protect globalvar from mutation in f()
  local globalvar="$globalvar"
  echo "$globalvar"
}

echo "$globalvar"
```

#### Here's what shellcheck currently says:
```
$ shellcheck myscript
No issues detected!

$
```

#### Here's what I wanted or expected to see:
```
$ shellcheck myscript
 
[Line 4:]
   local globalvar="$globalvar"
                    ^-- [SC2154](https://www.shellcheck.net/wiki/SC2154) (warning): globalvar is referenced but not assigned.

$
```
or at least something like:
```
$ shellcheck myscript
 
[Line 4:]
 local globalvar="$globalvar"
       ^-- [SC2269](https://www.shellcheck.net/wiki/SC2269) (info): This variable is assigned to itself, so the assignment does nothing.

$
```
or another rule like 
```
(warning): This variable is assigned to itself, so the assignment does nothing as there is no global variable assigned.
```
or
```
(warning): This variable is assigned to itself, so the assignment does nothing if there is no globally-available variable, check to ensure.
```

**Comment by wileyhy** [comment]
"protect globalvar from mutation in f()"

What does this mean precisely?

Prevent the creation of a local variable with the same name and different
value as a global variable?  Is the readonly attribute a part of the
equation? Is there an issue with variables which were properly declared as
local changing the values of global variables...?

Using a local copy of a global variable can have valid uses.


~ $ set -x; declare -p SHLVL SHELL
+ set -x
+ declare -p SHLVL SHELL
declare -x SHLVL="1"
declare -x SHELL="/data/data/com.termux/files/usr/bin/bash"
~ $ : $0
+ : /data/data/com.termux/files/usr/bin/bash
~ $ f(){ local SHELL=$( command -v ksh ); ksh; }
~ $ f
+ f
++ command -v ksh
+ local SHELL=/data/data/com.termux/files/usr/bin/ksh
+ ksh
localhost$ echo SHLVL:$SHLVL, SHELL:$SHELL, 0:$0
SHLVL:1, SHELL:/data/data/com.termux/files/usr/bin/ksh,
0:/data/data/com.termux/files/usr/bin/ksh
localhost$ exit
~ $ ksh
+ ksh
localhost$ echo SHELL:$SHELL
SHELL:/data/data/com.termux/files/usr/bin/bash
localhost$ ^D
~ $ : $SHELL
+ : /data/data/com.termux/files/usr/bin/bash
~ $


Wiley

**Comment by wileyhy** [comment]
Alternately...

# bash
~ $ declare -p IFS
declare -- IFS=$' \t\n'
~ $ IFS=' :'
~ $ g(){ local IFS="${IFS}|"; declare -p IFS; }
~ $ g
declare -- IFS=" :|"
~ $ declare -p IFS
declare -- IFS=" :"
~ $

**Comment by iconoclasthero** [comment]
> "protect globalvar from mutation in f()"
> 
> What does this mean precisely?
>  ⋮
> Using a local copy of a global variable can have valid uses.

I assume you're referring to the `## comment` in the example?  What it means is I want to make a `local` variable inside a function with the same name as the global variable while assigning the global value to it, e.g., as happens in the context of
```
globalvar=value
funcname(){
  local globalvar="$globalvar"
  ⋮
}
```
thus allowing the use of a variable of the same name (and value when the function is called) as a global variable but `local` to the function so as to "protect" the global from being changed/mutated by the function.  If, in the above, the first line, i.e., `globalvar=value` was absent and thus the global version of `$globalvar` did not exist when the local was created, SC2154 or something else like what I suggested in the OP be triggered.

I believe that we agree in principle that this is a useful property of bash, but what you're doing and what I'm doing are subtlety different in that when the local variable is created you're assigning a **new value** to the function local and I'm assigning the **current global value** to it and asking that if the global variable is not defined it get flagged by shellcheck.

I am aware that using a local copy of a global variable can have valid uses which is why I would like SC2154 to alert when doing so if the global version is not assigned.  Obviously, this begins to get complicated for shellcheck e.g., if the local value is mutated inside the function which declares the `local globalvar`, however, it would be helpful nonetheless.

**Comment by iconoclasthero** [comment]
I think the title of this issue is not particularly especially representative of the actual feature request (because I guess it's not exactly a bug?) so I'm open to suggestions for a better one.
